File: code/BloomFilter/BloomFilter.java
import java.io.BufferedWriter;
import java.io.IOException;
import java.util.BitSet;

public class BloomFilter {
	/*
	 * m = Number of Bits
	 * Bit
	 */
	private int m = 0;
	private BitSet container;
	
	/*
	 * k = Number of Hash Functions
	 */
	private int k;
	
	public BloomFilter(){
		//Default Parameters will have false positive rate of 1.0E-4 or (1 in 10,000) 
		//and is good for counting about 1M distinct values. 
		this(19170117, 13);
	}
	
	
	public BloomFilter(int bitSize, int numHash){
		this.m = bitSize;
		this.container = new BitSet(this.m);
		this.k = numHash;
	}
	
	public void add(Object obj){
		for(int idx: this.getBitIndexes(obj)){
			this.container.set(idx);
		}
	}
	
	
	
	
	public boolean contains(Object obj){
		for(int idx: this.getBitIndexes(obj)){
			if(this.container.get(idx) == false){
				return false;
			}
		}
		return true;
	}
	
	public  long getCount(){
		double numZeroBits = (double) this.m - this.container.cardinality();
		return (long) Math.round(Math.log(numZeroBits/this.m) / ((this.k * Math.log(1-1.0/this.m))));
	}
	
	public void merge(BloomFilter that) throws Exception{
		if(this.m != that.m){
			throw new Exception("Incompatible Bloom Filters. Bit size should be same");
		}
		if(this.k != that.k){
			throw new Exception("Incompatible Bloom Filters. Number of hash function should be same");
		}
		this.container.or(that.container);
	}
	
	
	private int[] getBitIndexes(Object obj){
		int[] indexes =  new int[this.k];
		for(int i=0; i<this.k; i++){
			indexes[i] =  Math.abs(MurmurHash.hash32(obj.toString() + new Integer(i).toString())) % this.m;
		}
		return indexes;
	}
	
	public byte[] getBytArray(){
		return this.container.toByteArray();
	}
	
	public void setContainer(byte[] input){
		this.container = BitSet.valueOf(input);
	}
	/*
	 * Source: Source: http://hur.st/bloomfilter?n=10000000&p=1.0E-6		
	 */
	public static int[] recommnededSize(long expectedItems, double expectedFP, int incrementBy){
		double n = new Long(expectedItems).doubleValue();
		int numOfBits = (int) Math.ceil((n * Math.log(expectedFP)) / Math.log(1.0 / (Math.pow(2.0, Math.log(2.0)))));
		int numOfHash = (int) Math.round(Math.log(2.0) * numOfBits/n);
		return new int[]{numOfBits, numOfHash};
	}
	
	public void fromString(String input){
		for(int i=0; i<this.m; i++){
			char c = input.charAt(i);
			if(c=='1'){ this.container.set(i);}			
		}		
	}
	
	public String toString(){
		StringBuffer output = new StringBuffer();
		for(int i=0; i<this.m; i++){
			boolean isSet = this.container.get(i);
			String text = (isSet ? "1" : "0");
			output.append(text);
		}		
		return output.toString();
	}
	
	

}

File: code/BloomFilter/BloomFilter.py
import math
import mmh3  #murmur hash function library

class BloomFilter(object):
    """Base Bloom Filter Class"""

    def __init__(self, m=None, k=None, n=None, fp=None):
        if not m == None and not k == None:
            self._k = k
            self._m = m
        elif not n == None and not fp == None:
            self._m, self._k = self._get_size(fp, n)
        else:
            raise Exception("Missing required parameters. Expecting either m and k or n and fp")

    def indexes(self, item):
        """Return Hash indexes for the given item"""
        values = []
        for i in range(self._k):
            itemStr = str(item) + str(i)
            value = mmh3.hash(itemStr) % self._m
            values.append(value)
        return values
        #return [int(mmh3.hash(str(item) + str(i)) 
        # % self._m) for i in range(self._k)]

    @property
    def count(self):
        #actual number of set bits
        t = float(self.set_bit_count)
        num = math.log(1 - t/self._m)
        deno = self._k * math.log(1 - 1.0/self._m)
        return round(num/deno)

    def _get_size(self, fp, n):
        """Returns bit size and number of hash functions to use for given n and fp"""

        m = int(math.ceil(-1 * (n * math.log(fp))/math.pow(math.log(2.0), 2.0)))
        k = int(math.ceil(m/n*math.log(2)))
        return [m, k]

    @property
    def container(self):
        return self._container

    @property
    def number_of_hashes(self):
        return self._k

    @property
    def bit_size(self):
        return self._m   

    @property
    def set_bit_count(self):
        pass     


class VectorBasedBloomFilter(BloomFilter):
    """Bloom Filter implementation where each bit is represented by vector index. Easier to understand what's going on but takes much more space"""

    def __init__(self, *args, **kwargs):
        super(VectorBasedBloomFilter, self).__init__(*args, **kwargs)
        self._container = [0] * self._m
        
    def add(self, item):
        for idx in self.indexes(item): 
            self._container[idx] = 1

    def contains(self, item):
        for idx in self.indexes(item): 
            if self._container[idx] == 0: 
                return  False
        return True

    def merge(self, other):
        if other == None: return
        
        if self.number_of_hashes != other.number_of_hashes:
            raise Exception("Number of hash function should be same in order to merge two bloom filters")
        if self.bit_size != other.bit_size:
            raise Exception("Bit size should be same in order to merge two bloom filters")

        for idx, value in enumerate(other.container):
            if value == 1:
                self._container[idx] = 1

    @property
    def set_bit_count(self):
        return sum(self._container)



class BitSetBasedBloomFilter(BloomFilter):
    """Bloom Filter implementation using Bitset -- much more space efficient as compared to VectorBasedBloomFilter"""

    def __init__(self, *args, **kwargs):
        super(BitSetBasedBloomFilter, self).__init__(*args, **kwargs)
        self._container = 0 << self._m+1
        

    def add(self, item):
        for idx in self.indexes(item): 
            self._container |= (1 << idx)

    def contains(self, item):
        for idx in self.indexes(item):
            if self._container & (1 << idx) == 0: return False
        return True

    def merge(self, other):
        if other == None: return
        
        if self.number_of_hashes != other.number_of_hashes:
            raise Exception("Number of hash function should be same in order to merge two bloom filters")
        if self.bit_size != other.bit_size:
            raise Exception("Bit size should be same in order to merge two bloom filters")

        self._container |= other.container


    @property
    def set_bit_count(self):
        return sum([1 for x in bin(self._container)[3:] if int(x)==1])


# class CountingBloomFilter(object):
#     def __init__(self, *args, **kwargs):
#         super(CountingBloomFilter, self).__init__(*args, **kwargs)

#         # build two dimensional structure for counting bloom filter
#         self._container = [[0] * self._m] * self._k   

#     def add(self, item):
#         for hashIdx, hashVal in enumerate(self.indexes(item)):
#             self._container[hashIdx][hashVal] += 1

#     def delete(self, item):
#         for hashIdx, hashVal in enumerate(self.indexes(item)):
#             if self._container[hashIdx][hashVal] > 0:
#                 self._container[hashIdx][hashVal] -= 1        

#     def item_count(self, item):
#         cnts = []
#         for hashIdx, hashVal in enumerate(self.indexes(item)):
#             cnts.append(self._container[hashIdx][hashVal])
#         return min(cnts)

#     @property
#     def count(self):





if __name__ == '__main__':
    bf1 = VectorBasedBloomFilter(fp=1E-4, n=10)
    bf1.add(2)
    bf1.add(6)
    bf1.add(3)
    bf1.add(2)
    print bf1.contains(3)
    print bf1.contains(10)
    print bf1.count

    # bf2 = VectorBasedBloomFilter(fp=1E-4, n=10)
    # bf2.add(5)
    # bf2.add(10)
    # bf2.add(2)
    # bf2.add(3)
    # print bf2.count

    # bf3 = VectorBasedBloomFilter(fp=1E-4, n=10)
    # bf3.merge(bf1)
    # bf3.merge(bf2)
    # print bf3.count

